File: bst.py
class Node:
    def __init__(self, val):
        self.val = val
        self.left = None
        self.right = None

class BST:
    def __init__(self):
        self.root = None

    def insert(self, val):
        if self.root is None:
            self.root = Node(val)
            return
        current = self.root
        while True:
            if val < current.val:
                if current.left is None:
                    current.left = Node(val)
                    return
                current = current.left
            else:
                if current.right is None:
                    current.right = Node(val)
                    return
                current = current.right

    def search(self, val):
        current = self.root
        while current:
            if val == current.val:
                return True
            elif val < current.val:
                current = current.left
            else:
                current = current.right
        return False

    def delete(self, val):
        self.root = self._delete(self.root, val)

    def _delete(self, node, val):
        if not node:
            return None
        if val < node.val:
            node.left = self._delete(node.left, val)
        elif val > node.val:
            node.right = self._delete(node.right, val)
        else:
            if not node.left:
                return node.right
            elif not node.right:
                return node.left
            successor = self._find_min(node.right)
            node.val = successor.val
            node.right = self._delete(node.right, successor.val)
        return node

    def _find_min(self, node):
        while node.left:
            node = node.left
        return node

    def in_order(self):
        result = []
        self._in_order(self.root, result)
        return result

    def _in_order(self, node, result):
        if node:
            self._in_order(node.left, result)
            result.append(node.val)
            self._in_order(node.right, result)

    def pre_order(self):
        result = []
        self._pre_order(self.root, result)
        return result

    def _pre_order(self, node, result):
        if node:
            result.append(node.val)
            self._pre_order(node.left, result)
            self._pre_order(node.right, result)

    def height(self):
        return self._height(self.root)

    def _height(self, node):
        if not node:
            return 0
        left_h = self._height(node.left)
        right_h = self._height(node.right)
        return max(left_h, right_h) + 1

if __name__ == "__main__":
    bst = BST()
    values = [50, 30, 70, 20, 40, 60, 80]
    print("Inserting values:", values)
    for v in values:
        bst.insert(v)
    print("\n── Traversals ──")
    print("In-Order  (sorted):", bst.in_order())
    print("Pre-Order         :", bst.pre_order())
    print("Tree Height       :", bst.height())
    print("\n── Search ──")
    print("Search(40) →", bst.search(40))
    print("Search(99) →", bst.search(99))
    print("\n── Delete 30 ──")
    bst.delete(30)
    print("In-Order after delete:", bst.in_order())

File: bst.go
package main

import "fmt"

type Node struct {
	Val   int
	Left  *Node
	Right *Node
}

type BST struct {
	Root *Node
}

func (b *BST) Insert(val int) {
	if b.Root == nil {
		b.Root = &Node{Val: val}
		return
	}
	current := b.Root
	for {
		if val < current.Val {
			if current.Left == nil {
				current.Left = &Node{Val: val}
				return
			}
			current = current.Left
		} else {
			if current.Right == nil {
				current.Right = &Node{Val: val}
				return
			}
			current = current.Right
		}
	}
}

func (b *BST) Search(val int) bool {
	current := b.Root
	for current != nil {
		if val == current.Val {
			return true
		} else if val < current.Val {
			current = current.Left
		} else {
			current = current.Right
		}
	}
	return false
}

func (b *BST) Delete(val int) {
	b.Root = deleteNode(b.Root, val)
}

func deleteNode(node *Node, val int) *Node {
	if node == nil {
		return nil
	}
	if val < node.Val {
		node.Left = deleteNode(node.Left, val)
	} else if val > node.Val {
		node.Right = deleteNode(node.Right, val)
	} else {
		if node.Left == nil {
			return node.Right
		} else if node.Right == nil {
			return node.Left
		}
		successor := findMin(node.Right)
		node.Val = successor.Val
		node.Right = deleteNode(node.Right, successor.Val)
	}
	return node
}

func findMin(node *Node) *Node {
	for node.Left != nil {
		node = node.Left
	}
	return node
}

func (b *BST) InOrder() []int {
	var result []int
	inOrder(b.Root, &result)
	return result
}

func inOrder(node *Node, result *[]int) {
	if node == nil {
		return
	}
	inOrder(node.Left, result)
	*result = append(*result, node.Val)
	inOrder(node.Right, result)
}

func (b *BST) PreOrder() []int {
	var result []int
	preOrder(b.Root, &result)
	return result
}

func preOrder(node *Node, result *[]int) {
	if node == nil {
		return
	}
	*result = append(*result, node.Val)
	preOrder(node.Left, result)
	preOrder(node.Right, result)
}

func (b *BST) Height() int {
	return height(b.Root)
}

func height(node *Node) int {
	if node == nil {
		return 0
	}
	leftH := height(node.Left)
	rightH := height(node.Right)
	if leftH > rightH {
		return leftH + 1
	}
	return rightH + 1
}

func main() {
	bst := &BST{}
	values := []int{50, 30, 70, 20, 40, 60, 80}
	fmt.Println("Inserting values:", values)
	for _, v := range values {
		bst.Insert(v)
	}
	fmt.Println("\n── Traversals ──")
	fmt.Println("In-Order  (sorted):", bst.InOrder())
	fmt.Println("Pre-Order         :", bst.PreOrder())
	fmt.Println("Tree Height       :", bst.Height())
	fmt.Println("\n── Search ──")
	fmt.Println("Search(40) →", bst.Search(40))
	fmt.Println("Search(99) →", bst.Search(99))
	fmt.Println("\n── Delete 30 ──")
	bst.Delete(30)
	fmt.Println("In-Order after delete:", bst.InOrder())
}